src/main/kotlin/uk/gov/justice/digital/hmpps/sentenceplan/services/GoalService.kt

```diff
@@ -49,15 +49,38 @@ class GoalService(
       areaOfNeed = areaOfNeedEntity,
       targetDate = goal.targetDate,
       plan = planEntity,
-      relatedAreasOfNeed = relatedAreasOfNeedEntity,
+      relatedAreasOfNeed = relatedAreasOfNeedEntity.toMutableList(),
       goalOrder = highestGoalOrder + 1,
-
     )
     val savedGoalEntity = goalRepository.save(goalEntity)
 
     return savedGoalEntity
   }
 
+  @Transactional
+  fun updateGoalByUuid(goalUuid: UUID, goal: Goal): GoalEntity {
+    val goalEntity = goalRepository.findByUuid(goalUuid)
+      ?: throw Exception("This Goal is not found: $goalUuid")
+
+    goalEntity.title = goal.title
+    goalEntity.targetDate = goal.targetDate
+
+    var relatedAreasOfNeedEntity: List<AreaOfNeedEntity> = emptyList()
+
+    if (goal.relatedAreasOfNeed.isNotEmpty()) {
+      relatedAreasOfNeedEntity = areaOfNeedRepository.findAllByNames(goal.relatedAreasOfNeed)
+        ?: throw Exception("One or more of the Related Areas of Need was not found: ${goal.relatedAreasOfNeed}")
+
+      if (goal.relatedAreasOfNeed.size != relatedAreasOfNeedEntity.size) {
+        throw Exception("One or more of the Related Areas of Need was not found")
+      }
```

Comment:
`findAllByNames` does a SELECT .. WHERE .. IN and so it will only return the Areas Of Need which explicitly match between the request and the database. If an API client passes in "Accommodation, Cheese" as its list of related areas of need names, it will return a single Area of Need object. I can't immediately think of a way of forcing the lookup to match on all incoming names and throw an exception if one of them doesn't match the DB.

Reply:
No I think it will only throw that exception if you pass it a list of AoNs where it doesn't recognise any of them. I think there might be a better version of findAllByName possible using `group by...having...` but I'd need to spend some time playing with it, and I'm not totally sure it's worth it.

Reply:
We could add a comment here explaining the behaviour if it is not clear?

Reply:
Fixed in https://github.com/ministryofjustice/hmpps-sentence-plan/pull/63/commits/445936be86f8b98ed61e578df5097b741c8558c7